### PR TITLE
Implement mem::{zeroed,uninitialized} in terms of MaybeUninit.

### PR DESCRIPTION
**Pull request by alex**
Refs #62061

r? @oli-obk

### CODE REVIEW

**Comment by oli-obk** [comment]
looks fine to me from a compiler perspective

maybe someone from @rust-lang/libs 	wants to look at the `MaybeUninit` usage?

cc @RalfJung

**Comment by RalfJung** [comment]
Fine from a `MybeUninit` perspective.

Let's ask perf though, just to be sure.

@bors try

**Comment by RalfJung** [comment]
Note that the `init` intrinsic is unused now, we we won't remove it yet as it will be used again in a follow-up PR.

**Comment by RalfJung** [comment]
Maybe it makes sense to mark the `init` intrinsic as deprecated already in this PR? Seems to be in-line with the other changes here.

**Comment by oli-obk** [comment]
good idea

**Comment by SimonSapin** [comment]
Why remove the calls to `panic_if_uninhabited`?

**Comment by Centril** [comment]
@SimonSapin They happen in `.assume_init()`.

**Comment by alex** [comment]
Hmm, simply sticking a `#[rustc_deprecated()]` on the `fn init` results in an error that `rustc_deprecated attribute must be paired with either stable or unstable attribute`. The whole module is marked `unstable` though. Is there an appropriate idiom here? Simply mark the function `unstable` again?

**Comment by Centril** [comment]
r? @RalfJung 

Feel free to review this since the PR doesn't change anything observably.

**Comment by SimonSapin** [comment]
@alex Yes, items without a `#[stable]` or `#[unstable]` attribute “inherit” that of their parent item. (At least when the parent is a module.) So duplicating that attribute to satisfy the lint does not change the stability of the function.

**Comment by alex** [comment]
Oof. Is the correct fix just to `#![allow(deprecated)` for that test?

**Comment by Centril** [comment]
I think we should do the steps @oli-obk outlined in https://github.com/rust-lang/rust/issues/62061#issuecomment-505851241. If we do that, then imo we can just remove `intrinsics::init` here and now.

**Comment by RalfJung** [comment]
That seems like a change with enough potential for problems that I wouldn't want to block removing `uninit` on it.

> Is the correct fix just to #![allow(deprecated) for that test?

Yes.

**Comment by oli-obk** [comment]
> That seems like a change with enough potential for problems that I wouldn't want to block removing uninit on it.

I think the idea was not to block this PR, but to just nix `init` here, too and think about future steps in separate PRs.

**Comment by RalfJung** [comment]
Well that would be fine for me. It would commit ourselves to not follow [this plan](https://github.com/rust-lang/rust/issues/62061#issuecomment-505813837) and instead do the "proper" thing right away. I just wanted to not stand in the way of making `MaybeUninit::zeroed` a `const fn` ASAP since some people have now waited for it for more than half a year.

But I will never argue against nixing an intrinsic such as `init`. ;)

**Comment by alex** [comment]
Ok, should be green now.

I don't know nearly enough about the `&mut` in `const fn` issue to take that on unfortunately.

**Comment by RalfJung** [comment]
LGTM, unless @Centril and @oli-obk want to push through with the idea of nixing `init` as well?

We could easily do that in a separate PR though.

**Comment by Centril** [comment]
I'm totally fine with a separate PR :)

@bors r=RalfJung,oli-obk,Centril rollup

**Comment by Centril** [comment]
Let's try this again.

@bors r=RalfJung,oli-obk,Centril rollup

**Comment by RalfJung** [comment]
That would be

@bors ping

I think. Cc @rust-lang/infra

**Comment by RalfJung** [comment]
Looks like that CI update has grinded rustc to a halt for >24h now.

**Comment by Centril** [comment]
Failed (clippy toolstate regression) in https://github.com/rust-lang/rust/pull/62182#issuecomment-506507297, @bors r-

I think you can fix it by removing the deprecation of the intrinsic for now and deferring that to another PR.

**Comment by alex** [comment]
Removed the deprecation (and the other changes resulting from it).

**Comment by RalfJung** [comment]
Could you add a `FIXME: deprecate once clippy's tests are fixed` or so?

**Comment by alex** [comment]
@RalfJung done!

**Comment by RalfJung** [comment]
Thanks!

@bors r=RalfJung,oli-obk,Centril

**Comment by alex** [comment]
So it seems this would break the clippy tests -- they have a test involving `std::intrinsics::uninit`. How does this work, do I send them a PR first, or can clippy's tests just be broken on master?

**Comment by oli-obk** [comment]
In this case sending a PR to clippy first is doable and would be best

**Comment by RalfJung** [comment]
You can remove the `init` test at the same time, or add an `allow(deprecated)` there, to prepare for that.

**Comment by RalfJung** [comment]
That gives us enough time to do a perf run...

@rust-timer build c6c5b77f3f51db970b8e506e00f052c6254cbe4a

**Comment by alex** [comment]
Alrighty https://github.com/rust-lang/rust-clippy/pull/4240 is up -- I assume once this lands I'll need to send a rustc PR to bump the clippy submodule.

**Comment by RalfJung** [comment]
Yes, or you can do that as part of this PR.

**Comment by Centril** [comment]
Perf looks good enough / noise imo.

**Comment by alex** [comment]
Clippy side landed, https://github.com/rust-lang/rust/pull/62246 is up now.

File: src/libcore/intrinsics.rs [hunk 1]
```diff
@@ -700,17 +700,9 @@ extern "rust-intrinsic" {
     /// which is unsafe unless `T` is `Copy`. Also, even if T is
     /// `Copy`, an all-zero value may not correspond to any legitimate
     /// state for the type in question.
+    // FIXME: deprecate once clippy's tests are fixed
```
Comment (RalfJung) [inline src/libcore/intrinsics.rs]: The "no tools shall break" freeze is basically over... I think you can just deprecate this, and we can fix clippy asynchronously.

Comment (alex) [reply src/libcore/intrinsics.rs]: Will do once #62246 is merged

Comment (RalfJung) [reply src/libcore/intrinsics.rs]: Why wait? Breaking clippy *tests* brielfy is not very uncommon.

The reasong it didn't work before is that we have a one-week "freeze window" before the beta branches, where tools may not regress.

Comment (alex) [reply src/libcore/intrinsics.rs]: Ah, gotcha. Done!

**Comment by RalfJung** [comment]
Thanks!

@bors r+

**Comment by RalfJung** [comment]
Seems like this PR caused SIGILL in a bunch of libraries (see the backlinks GH inserted above), but it also seems like these libraries all had UB -- so tiny changes in the implementations of these methods can lead to arbitrary changes in behavior.

**Comment by ishitatsuyuki** [comment]
Based on what described in https://github.com/amethyst/amethyst/issues/1808#issuecomment-513449537, I suggest that this change to be backed out for at least one stable release. This kind of UB existed in old dependencies that are no longer maintained, therefore some crates have to migrate their dependency to the latest version to mitigate the issue, and we need some time for the ecosystem to do that major version upgrades.

**Comment by RalfJung** [comment]
@ishitatsuyuki a closed PR is not a good place for a new discussion. Could you open a new issue? Then I will nominate it for lang team discussion.

**Pull request merged**